### PR TITLE
[ncp] integrate InfraIf module with NcpHost to make border routing work

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR integrates the `InfraIf` module with `NcpHost` to make the border routing function work under NCP mode.

The PR also adds an integration test to test the border routing functionality. To execute this test, docker is used to simulate two devices connected by a infrastructure link (docker network bridge). otbr-agent runs inside the docker and we ping from the host machine to a simulated cli ftd nodes using its OMR address. Some code is added in the test scripts to support testing with docker.

### CODE REVIEW

**Comment by Irving-cl** [comment]
This PR depends on https://github.com/openthread/ot-br-posix/pull/2563.

File: src/ncp/ncp_spinel.cpp [hunk 1]
```diff
@@ -413,6 +413,18 @@ void NcpSpinel::HandleValueIs(spinel_prop_key_t aKey, const uint8_t *aBuffer, ui
...
     }
 
+    case SPINEL_PROP_INFRA_IF_SEND_ICMP6:
+    {
+        uint32_t            infraIfIndex;
+        const otIp6Address *destAddress;
+        const uint8_t      *data;
+        uint16_t            dataLen;
+
+        SuccessOrExit(ParseInfraIfIcmp6Nd(aBuffer, aLength, infraIfIndex, destAddress, data, dataLen));
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:423]: The error is discarded.

File: src/ncp/ncp_spinel.cpp [hunk 2]
```diff
@@ -751,6 +774,72 @@ otError NcpSpinel::ParseOperationalDatasetTlvs(const uint8_t            *aBuf,
...
+    otbrError    error        = OTBR_ERROR_NONE;
+    EncodingFunc encodingFunc = [this, aInfraIfIndex, aIsRunning, &aIp6Addresses] {
+        otError error = OT_ERROR_NONE;
+        SuccessOrExit(error = mEncoder.WriteUint32(aInfraIfIndex));
+        SuccessOrExit(error = mEncoder.WriteBool(aIsRunning));
+        for (const Ip6Address &addr : aIp6Addresses)
+        {
+            SuccessOrExit(error = mEncoder.WriteIp6Address(reinterpret_cast<const otIp6Address &>(addr)));
+        }
+    exit:
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:809]:
```suggestion

    exit:
```

File: src/ncp/ncp_spinel.cpp [hunk 3]
```diff
@@ -751,6 +774,72 @@ otError NcpSpinel::ParseOperationalDatasetTlvs(const uint8_t            *aBuf,
...
+        return error;
+    };
+
+    SuccessOrExit(SetProperty(SPINEL_PROP_INFRA_IF_STATE, encodingFunc), error = OTBR_ERROR_OPENTHREAD);
+exit:
```
Comment (superwhd) [inline src/ncp/ncp_spinel.cpp:814]:
```suggestion

exit:
```

**Review by superwhd — APPROVED** [review]
LGTM with minor suggestions

**Review by bukepo — APPROVED** [review]

**Review by bukepo — APPROVED** [review]

**Review by abtink — APPROVED** [review]
Looks great.

**Review by jwhui — APPROVED** [review]

**Pull request merged**